### PR TITLE
Replace internal metrics lib with Prometheus client. 

### PR DESCRIPTION
**Pull request by bergquist**
ref #6742

- [x] implement graphite bridge that support deltas
- [x] replace internal metriclib with prometheus
- [x] update alert executor to use new prometheus client
- [x] group metrics based of first part of metric name in graphite.
- [x] improve how settings are read in grafana-server. 
- [x] make sure scrapes does not create sessions.
- [x] add metrics that wont be sent to graphite due to amount of tags. ex http_request_total
- [x] make prefix backwards compatible
- [x] docs
- [x] add handler label thats supported by graphite and prometheus. rename the "code" tag to "status-code" to have correct order of labels in graphite.

### CODE REVIEW

**Comment by RichiH** [comment]
@brian-brazil @superq :)

**Comment by bergquist** [comment]
FYI. the bridge is a bit hacky but It will do for now. 

All metrics can be found in https://github.com/grafana/grafana/blob/poc_internal_metrics/pkg/metrics/metrics.go#L60 

Suggestions for new metrics/labels for current metrics would be great.

File: pkg/middleware/request_metrics.go [hunk 1]
```diff
@@ -28,38 +36,165 @@ func RequestMetrics() macaron.Handler {
...
+		return "notify"
+	default:
+		return strings.ToLower(m)
+	}
+}
+
+// If the wrapped http.Handler has not set a status code, i.e. the value is
+// currently 0, santizeCode will return 200, for consistency with behavior in
+// the stdlib.
+func sanitizeCode(s int) string {
```
Comment (SuperQ) [inline pkg/middleware/request_metrics.go]: In this one, the only case you need is the `0` to `200` mappings.  All of the rest can be handled by the default case.

File: pkg/middleware/request_metrics.go [hunk 2]
```diff
@@ -28,38 +36,165 @@ func RequestMetrics() macaron.Handler {
...
+		metrics.M_Proxy_Status.WithLabelValues("400").Inc()
 	case 500:
-		metrics.M_Proxy_Status_500.Inc(1)
+		metrics.M_Proxy_Status.WithLabelValues("500").Inc()
+	default:
+		metrics.M_Proxy_Status.WithLabelValues("unknown").Inc()
+	}
+}
+
+func sanitizeMethod(m string) string {
```
Comment (bergquist) [inline pkg/middleware/request_metrics.go]: True. I just stole it from https://github.com/prometheus/client_golang/blob/master/prometheus/promhttp/instrument_server.go#L60

I guess its for performance reason.

File: pkg/middleware/request_metrics.go (hunk 1 above)
Comment (bergquist) [reply pkg/middleware/request_metrics.go]: Same as above.

File: pkg/middleware/request_metrics.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (bergquist) [reply pkg/middleware/request_metrics.go]: True. I just stole it from https://github.com/prometheus/client_golang/blob/master/prometheus/promhttp/instrument_server.go#L60

I guess its for performance reason.

[the remark this answers is not in the document]
Comment (SuperQ) [reply pkg/middleware/request_metrics.go]: Ahh, interesting optimization.

**Comment by bergquist** [comment]
@RichiH @SuperQ 
I refactored our route registration a bit to make the route available as a label in prometheus.
Scrapes now look like this https://gist.githubusercontent.com/bergquist/00811747655e424285e1fa55df2322cf/raw/5a3557cc9ae6e7e63033279ee86d9865f49d06da/metrics%2520scrape does it make sense?

**Comment by RichiH** [comment]
Please prefix grafana_.

Else, I like it better.

Richard

Sent by mobile; excuse my brevity and the wall of text Gmail appends by
default.

**Comment by SuperQ** [comment]
I think @RichiH is talking about prefixing the custom metrics with `grafana_`.  For example `api_...` and `alerting_...`.

**Comment by RichiH** [comment]
> I think @RichiH is talking about prefixing the custom metrics with grafana_. For example api_... and alerting_....

He is.

**Comment by bergquist** [comment]
Fixed https://gist.githubusercontent.com/bergquist/00811747655e424285e1fa55df2322cf/raw/1044ae71245e9df70fe96ea1789dcc4d1650b7a8/metrics%2520scrape

Now I just need to cleanup this branch before merging.

File: pkg/metrics/metrics.go [hunk 3]
```diff
@@ -58,29 +58,29 @@ var (
 
 func init() {
 	M_Instance_Start = prometheus.NewCounter(prometheus.CounterOpts{
-		Name: "instance_start_total",
+		Name: "grafana_instance_start_total",
```
Comment (SuperQ) [inline pkg/metrics/metrics.go]: Typically we use the Prometheus `Namespace` in the [Opts](https://godoc.org/github.com/prometheus/client_golang/prometheus#Opts) struct.

Here's another example: 
https://github.com/prometheus/mysqld_exporter/blob/master/collector/exporter.go#L78-L83

**Comment by bergquist** [comment]
rebased onto master and squashed some commits.

**Pull request merged**